pkg/feed/datatracker.go

```diff
@@ -0,0 +1,192 @@
+	if sharedIndexes == nil {
+		sharedIndexes = make(map[string]any)
+	}
+
+	sourceID := source.GetIdentifier()
+
+	// Check if we have a cached index for this source
+	var sharedIndex *VersionedIndex[T]
+	if cached, exists := sharedIndexes[sourceID]; exists {
+		if typedIndex, ok := cached.(*VersionedIndex[T]); ok {
+			sharedIndex = typedIndex
+		}
+	}
+
+	// If version is zero value, get the current version
+	var zeroT T
+	if version == zeroT {
+		var err error
+		version, err = source.GetCurrentVersion(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "getting current version")
+		}
+	}
+
+	// If we don't yet have the in-memory cache, or if version changed, read the cache from the data source.
+	if sharedIndex == nil || version != sharedIndex.Version {
+		ready := make(chan struct{})
+		sharedIndex = &VersionedIndex[T]{Version: version, Ready: ready}
+		sharedIndexes[sourceID] = sharedIndex
+
+		go func() {
+			defer close(ready)
+
+			log.Printf("Fetching tracked packages list from %s version %v...\n", source.GetIdentifier(), version)
+
+			if err := loadIndex(ctx, source, version); err != nil {
+				sharedIndex.Err = err
+				return
+			}
+		}()
+	}
+
+	return TrackerFromFunc(func(e schema.TargetEvent) (bool, error) {
+		<-sharedIndex.Ready
+		if sharedIndex.Err != nil {
+			return false, sharedIndex.Err
+		}
+
+		if version != sharedIndex.Version {
+			log.Printf("Warning: the index version has changed since initialization (expected %v, got %v)",
+				version, sharedIndex.Version)
+		}
+
+		if _, ok := sharedIndex.Index[e.Ecosystem]; !ok {
+			return false, nil
+		}
+
+		tracked, ok := sharedIndex.Index[e.Ecosystem][e.Package]
+		return ok && tracked, nil
+	}), nil
+}
+
+// loadIndex loads the tracked packages index from the data source
+func loadIndex[T comparable](ctx context.Context, source DataSource[T], version T) error {
+	sourceID := source.GetIdentifier()
+	sharedIndex := sharedIndexes[sourceID].(*VersionedIndex[T])
+
+	r, err := source.GetReader(ctx, version)
+	if err != nil {
+		return errors.Wrap(err, "opening data source reader")
+	}
+	defer logFailure(r.Close)
+
+	gzr, err := gzip.NewReader(r)
+	if err != nil {
+		return errors.Wrap(err, "opening gzip reader")
+	}
+	defer logFailure(gzr.Close)
+
+	var rawTracked TrackedPackageSet
+	if err := json.NewDecoder(gzr).Decode(&rawTracked); err != nil {
+		return errors.Wrap(err, "unmarshalling tracked packages file")
+	}
+
+	trackedPackages := make(TrackedPackageIndex)
+	var totalPackages int
+	for eco, pkgs := range rawTracked {
+		if _, ok := trackedPackages[eco]; !ok {
+			trackedPackages[eco] = make(map[string]bool)
+		}
+		for _, pkgName := range pkgs {
+			trackedPackages[eco][pkgName] = true
+		}
+		totalPackages += len(pkgs)
+	}
+
+	log.Printf("Loaded index of %d tracked packages\n", totalPackages)
+	sharedIndex.Index = trackedPackages
+	return nil
+}
+
+// GCSDataSource implements DataSource for Google Cloud Storage using int64 generations
+type GCSDataSource struct {
```

Comment:
nit: Want to call this GCSObjectDataSource? This is specific to a single object and narrowing the naming allows for potentially other versioning strategies built on GCS.